File: lib/mpl_bubblegum_ex/metadata.ex
defmodule MplBubblegumEx.Creator do
  @moduledoc """
  Represents a creator of an NFT.
  """
  defstruct [
    :address,    # Base58 encoded Solana public key
    :verified,   # Boolean indicating if the creator is verified
    :share       # Percentage share (0-100) of royalties
  ]

  @type t :: %__MODULE__{
    address: String.t(),
    verified: boolean(),
    share: integer()
  }
end

defmodule MplBubblegumEx.Collection do
  @moduledoc """
  Represents a collection that an NFT belongs to.
  """
  defstruct [
    :verified,   # Boolean indicating if the collection is verified
    :key         # Base58 encoded Solana public key of the collection
  ]

  @type t :: %__MODULE__{
    verified: boolean(),
    key: String.t()
  }
end

defmodule MplBubblegumEx.Uses do
  @moduledoc """
  Represents the usage configuration for an NFT.
  """
  defstruct [
    :use_method,   # Atom: :burn, :multiple, or :single
    :remaining,    # Integer representing remaining uses
    :total         # Integer representing total uses
  ]

  @type use_method :: :burn | :multiple | :single
  @type t :: %__MODULE__{
    use_method: use_method(),
    remaining: non_neg_integer(),
    total: non_neg_integer()
  }
end

defmodule MplBubblegumEx.TokenStandard do
  @type t :: :non_fungible | :fungible_asset | :fungible | :non_fungible_edition
end

defmodule MplBubblegumEx.MetadataArgs do
  @moduledoc """
  Represents the metadata for a compressed NFT.
  """
  defstruct [
    :name,                    # Name of the NFT
    :symbol,                  # Symbol/ticker for the NFT
    :uri,                     # URI to the JSON metadata
    :seller_fee_basis_points, # Royalty fee in basis points (100 = 1%)
    :primary_sale_happened,   # Whether the primary sale has happened
    :is_mutable,              # Whether the NFT can be updated after minting
    :edition_nonce,           # Edition nonce (optional)
    :token_standard,          # Token standard (optional)
    :collection,              # Collection info (optional)
    :uses,                    # Uses info (optional)
    :creators                 # List of creators
  ]

  @type t :: %__MODULE__{
    name: String.t(),
    symbol: String.t(),
    uri: String.t(),
    seller_fee_basis_points: integer(),
    primary_sale_happened: boolean(),
    is_mutable: boolean(),
    edition_nonce: integer() | nil,
    token_standard: MplBubblegumEx.TokenStandard.t() | nil,
    collection: MplBubblegumEx.Collection.t() | nil,
    uses: MplBubblegumEx.Uses.t() | nil,
    creators: [MplBubblegumEx.Creator.t()]
  }

  @doc """
  Creates a new MetadataArgs struct with sane defaults.
  """
  def new(name, uri, creators \\ []) do
    %__MODULE__{
      name: name,
      symbol: "",
      uri: uri,
      seller_fee_basis_points: 0,
      primary_sale_happened: false,
      is_mutable: true,
      edition_nonce: nil,
      token_standard: :non_fungible,
      collection: nil,
      uses: nil,
      creators: creators
    }
  end

  @doc """
  Validates a metadata struct.
  """
  def validate(%__MODULE__{} = metadata) do
    with :ok <- validate_name(metadata.name),
         :ok <- validate_uri(metadata.uri),
         :ok <- validate_creators(metadata.creators) do
      {:ok, metadata}
    end
  end

  defp validate_name(name) when byte_size(name) > 0 and byte_size(name) <= 32, do: :ok
  defp validate_name(_), do: {:error, "Name must be between 1 and 32 bytes"}

  defp validate_uri(uri) when byte_size(uri) > 0 and byte_size(uri) <= 200, do: :ok
  defp validate_uri(_), do: {:error, "URI must be between 1 and 200 bytes"}

  defp validate_creators([]), do: :ok
  defp validate_creators(creators) do
    # Check if all creators have valid addresses
    invalid_addresses = Enum.filter(creators, fn c ->
      !MplBubblegumEx.Native.validate_pubkey_nif(c.address)
    end)

    if length(invalid_addresses) > 0 do
      {:error, "Invalid creator addresses: #{inspect(Enum.map(invalid_addresses, & &1.address))}"}
    else
      # Check if shares add up to 100
      total_shares = Enum.reduce(creators, 0, fn c, acc -> acc + c.share end)
      if total_shares != 100, do: {:error, "Creator shares must sum to 100, got #{total_shares}"}, else: :ok
    end
  end
end

defmodule MplBubblegumEx.MintOptions do
  @moduledoc """
  Options for minting NFTs.
  """
  defstruct [
    :mint_to_collection,              # Boolean - whether to mint to a collection
    :collection_key,                  # Base58 pubkey of the collection
    :collection_authority,            # Binary - keypair of collection authority
    :collection_authority_record_pda  # Base58 pubkey of authority record (if delegated)
  ]

  @type t :: %__MODULE__{
    mint_to_collection: boolean() | nil,
    collection_key: String.t() | nil,
    collection_authority: binary() | nil,
    collection_authority_record_pda: String.t() | nil
  }
end

File: native/mpl_bubblegum_native/src/metadata.rs
use mpl_bubblegum::types::{
    Creator, MetadataArgs, TokenProgramVersion, TokenStandard, Collection, Uses, UseMethod
};
use rustler::{NifStruct, NifUnitEnum};
use solana_program::pubkey::Pubkey;
use std::str::FromStr;

// NIF struct to receive creator data from Elixir
#[derive(NifStruct)]
#[module = "MplBubblegumEx.Creator"]
pub struct CreatorNif {
    pub address: String,
    pub verified: bool,
    pub share: u8,
}

// NIF struct to receive collection data from Elixir
#[derive(NifStruct)]
#[module = "MplBubblegumEx.Collection"]
pub struct CollectionNif {
    pub verified: bool,
    pub key: String,
}

// NIF enum for token standard
#[derive(NifUnitEnum)]
pub enum TokenStandardNif {
    NonFungible,
    FungibleAsset,
    Fungible,
    NonFungibleEdition,
}

// NIF struct to receive use data from Elixir
#[derive(NifStruct)]
#[module = "MplBubblegumEx.Uses"]
pub struct UsesNif {
    pub use_method: UseMethodNif,
    pub remaining: u64,
    pub total: u64,
}

// NIF enum for use method
#[derive(NifUnitEnum)]
pub enum UseMethodNif {
    Burn,
    Multiple,
    Single,
}

// NIF struct for metadata args from Elixir
#[derive(NifStruct)]
#[module = "MplBubblegumEx.MetadataArgs"]
pub struct MetadataArgsNif {
    pub name: String,   
    pub symbol: String,
    pub uri: String,
    pub seller_fee_basis_points: u16,
    pub primary_sale_happened: bool,
    pub is_mutable: bool,
    pub edition_nonce: Option<u8>,
    pub token_standard: Option<TokenStandardNif>,
    pub collection: Option<CollectionNif>,
    pub uses: Option<UsesNif>,
    pub creators: Vec<CreatorNif>,
}

// Convert Elixir metadata to mpl-bubblegum MetadataArgs
pub fn convert_metadata_args(args: &MetadataArgsNif) -> Result<MetadataArgs, String> {
    // Convert creators
    let creators: Vec<Creator> = args.creators
        .iter()
        .map(|c| {
            let address = match Pubkey::from_str(&c.address) {
                Ok(pubkey) => pubkey,
                Err(_) => return Err(format!("Invalid creator address: {}", c.address)),
            };
            
            Ok(Creator {
                address,
                verified: c.verified,
                share: c.share,
            })
        })
        .collect::<Result<Vec<Creator>, String>>()?;
    
    // Validate creator shares
    let total_shares: u16 = creators.iter().map(|c| c.share as u16).sum();
    if !creators.is_empty() && total_shares != 100 {
        return Err(format!("Creator shares must sum to 100, got {}", total_shares));
    }
    
    // Convert collection if present
    let collection = args.collection.as_ref().map(|c| {
        let key = match Pubkey::from_str(&c.key) {
            Ok(pubkey) => pubkey,
            Err(_) => return Err(format!("Invalid collection key: {}", c.key)),
        };
        
        Ok(Collection {
            verified: c.verified,
            key,
        })
    }).transpose()?;
    
    // Convert uses if present
    let uses = args.uses.as_ref().map(|u| {
        let use_method = match u.use_method {
            UseMethodNif::Burn => UseMethod::Burn,
            UseMethodNif::Multiple => UseMethod::Multiple,
            UseMethodNif::Single => UseMethod::Single,
        };
        
        Uses {
            use_method,
            remaining: u64::try_from(u.remaining).unwrap(),
            total: u64::try_from(u.total).unwrap(),
        }
    });
    
    // Convert token standard if present
    let token_standard = args.token_standard.as_ref().map(|ts| {
        match ts {
            TokenStandardNif::NonFungible => TokenStandard::NonFungible,
            TokenStandardNif::FungibleAsset => TokenStandard::FungibleAsset,
            TokenStandardNif::Fungible => TokenStandard::Fungible,
            TokenStandardNif::NonFungibleEdition => TokenStandard::NonFungibleEdition,
        }
    });
    
    Ok(MetadataArgs {
        name: args.name.clone(),
        symbol: args.symbol.clone(),
        uri: args.uri.clone(),
        seller_fee_basis_points: args.seller_fee_basis_points,
        primary_sale_happened: args.primary_sale_happened,
        is_mutable: args.is_mutable,
        edition_nonce: args.edition_nonce,
        token_standard,
        collection,
        uses,
        token_program_version:TokenProgramVersion::Original,
        creators,
    })
}
